randperm and arange op support S0

grybd opened the pull request:
这个PR的目的
- [x] randperm op 支持S0 、添加单测
- [x] arange op支持 S0、添加单测

grybd commented:
在处理source op 支持S0中遇到的问题
#### 问题一
randint op 在 graph中进行S0切分后，在consisitent视角下总是对称的，但是在eager模式下是非对称的，感觉有些困惑
例如：
```python
def _test_consistent_rand_graph(low, high, shape, placement):
    class ConsistentRandGraph(flow.nn.Graph):
        def __init__(self,):
            super().__init__()

        def build(self):
            x = flow.randint(low, high, shape, placement=placement, sbp=flow.sbp.broadcast)
            y = flow.randint(low,high,shape,placement=placement,sbp=flow.sbp.split(0))
            return [x,y]

    c_r_g = ConsistentRandGraph()
    res = c_r_g()
    print("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
    print("x: ",res[0])
    print("x.to_local(): ",res[0].to_local())
    print("-----------------------------------------------------------")
    print("y: ",res[1])
    print("y to_local(): ",res[1].to_local())

if __name__ == "__main__":
    _test_consistent_rand_graph(1,10 ,(10,), placement=flow.env.all_device_placement("cuda"))
    _test_consistent_rand_graph(1,10 ,(10,), placement=flow.env.all_device_placement("cuda"))
    
```
执行 ： ```shell   python3 -m oneflow.distributed.launch --nproc_per_node 2 --master_port 12345 /home/fengdaochao/fdz/test.py ```

输出 ：
```shell
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x: +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ 
x:  tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)

x.to_local(): x.to_local():   tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4], device='cuda:0', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4], device='cuda:1', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([3, 7, 7, 5, 1, 3, 7, 7, 5, 1],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([3, 7, 7, 5, 1, 3, 7, 7, 5, 1],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([3, 7, 7, 5, 1], device='cuda:0', dtype=oneflow.int64)
tensor([3, 7, 7, 5, 1], device='cuda:1', dtype=oneflow.int64)
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4], device='cuda:1', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4], device='cuda:0', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([3, 7, 7, 5, 1, 3, 7, 7, 5, 1],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([3, 7, 7, 5, 1, 3, 7, 7, 5, 1],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([3, 7, 7, 5, 1], device='cuda:1', dtype=oneflow.int64)
tensor([3, 7, 7, 5, 1], device='cuda:0', dtype=oneflow.int64) ```

可以看到输出y tensor中间发生了由 “y:  tensor([3, 7, 7, 5, 1, 4, 4, 2, 5, 4], device='cuda:0', dtype=oneflow.int64)” ==》“y:  tensor([3, 7, 7, 5, 1, 3, 7, 7, 5, 1],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)” 这是合理的吗？，这样的转换是不是会丢失一些数据导致运算不正确

在一个终端输出所有信息不太好看，用两个终端来显示输出结果
T0 输出：
```shell
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6], device='cuda:0', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([7, 2, 7, 3, 6, 7, 2, 7, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([7, 2, 7, 3, 6], device='cuda:0', dtype=oneflow.int64)
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6], device='cuda:0', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([7, 2, 7, 3, 6, 7, 2, 7, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([7, 2, 7, 3, 6], device='cuda:0', dtype=oneflow.int64) ```

T1 输出：
```shell
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6], device='cuda:1', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([7, 2, 7, 3, 6, 7, 2, 7, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([7, 2, 7, 3, 6], device='cuda:1', dtype=oneflow.int64)
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([7, 2, 7, 3, 6, 5, 2, 1, 3, 6], device='cuda:1', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([7, 2, 7, 3, 6, 7, 2, 7, 3, 6],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y to_local():  tensor([7, 2, 7, 3, 6], device='cuda:1', dtype=oneflow.int64) ```

grybd commented:
test_arange.py 也可以跑通了

grybd commented:
在处理source op 支持S0中遇到的问题
randint op 在 graph中进行S0切分后输出tensor在consisten视角下总是对称的，但是在eager模式下输出结果是随机、非对称的，初步判断应该是rank0 、rank1中拿到了同样的seed 然后根据S0推导出的shape产生tensor，这样拼接起来就是对称的，但这样的处理方法应该是不行的，会丢失部分数据(randperm op和arange op也存在这个问题)
例如：
```python
def _test_consistent_rand_graph(low, high, shape, placement):
    class ConsistentRandGraph(flow.nn.Graph):
        def __init__(self,):
            super().__init__()

        def build(self):
            x = flow.randint(low, high, shape, placement=placement, sbp=flow.sbp.broadcast)
            y = flow.randint(low,high,shape,placement=placement,sbp=flow.sbp.split(0))
            return [x,y]

    c_r_g = ConsistentRandGraph()
    res = c_r_g()
    print("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
    print("x: ",res[0])
    print("x.to_local(): ",res[0].to_local())
    print("-----------------------------------------------------------")
    print("y: ",res[1])
    print("y.to_local(): ",res[1].to_local())

if __name__ == "__main__":
    _test_consistent_rand_graph(1,10 ,(10,), placement=flow.env.all_device_placement("cuda"))
    print("============================================================")
    print("eager mode: ",flow.randint(1,10,(10,),placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.split(0)))
    print("============================================================")
    
```
在T0和T1终端分别执行 命令：```shell python3 /home/fengdaochao/fdz/test.py ```
T0 输出：
```shell
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  tensor([5, 4, 9, 1, 7, 6, 1, 8, 6, 7],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([5, 4, 9, 1, 7, 6, 1, 8, 6, 7], device='cuda:0', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([5, 4, 9, 1, 7, 5, 4, 9, 1, 7],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y.to_local():  tensor([5, 4, 9, 1, 7], device='cuda:0', dtype=oneflow.int64)
============================================================
eager mode:  tensor([5, 4, 9, 1, 7, 1, 8, 4, 4, 2],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
============================================================
 ```
T1 输出：
```shell
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
x:  tensor([5, 4, 9, 1, 7, 6, 1, 8, 6, 7],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.broadcast,), dtype=oneflow.int64)
x.to_local():  tensor([5, 4, 9, 1, 7, 6, 1, 8, 6, 7], device='cuda:1', dtype=oneflow.int64)
-----------------------------------------------------------
y:  tensor([5, 4, 9, 1, 7, 5, 4, 9, 1, 7],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y.to_local():  tensor([5, 4, 9, 1, 7], device='cuda:1', dtype=oneflow.int64)
============================================================
eager mode:  tensor([5, 4, 9, 1, 7, 1, 8, 4, 4, 2],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
============================================================
 ```

可以看到eager mode输出结果tensor([5, 4, 9, 1, 7, 1, 8, 4, 4, 2],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)是随机的非对称的,而在module里调用 randint op进行S0 的输出tensor y:  tensor([5, 4, 9, 1, 7, 5, 4, 9, 1, 7],
       placement=oneflow.placement(device_type="cuda", device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64) 是对称的 ，这样推导shape没问题,但是这样会丢失一部分随机数据

arange op 和 randperm op也同样存在这样的问题

strint commented:
1、补上大致的调用路径，pyhon api -> c++ generator的调用
2、约一个讨论，参见同事 chengbowen，chengcheng

strint commented:
```
__global__ void InitCurandStatesKernel(uint64_t seed, curandState* states,
                                       CUDAGeneratorState* cuda_gen_state) {
  const int id = blockIdx.x * blockDim.x + threadIdx.x;
  size_t local_seed = (static_cast<size_t>(seed) + 0x9e3779b9U + (static_cast<size_t>(id) << 6U)
                       + (static_cast<size_t>(id) >> 2U));
  curand_init(local_seed, 0, 0, &states[id]);
  cuda_gen_state->dev_counter = static_cast<int32_t>(0);
  cuda_gen_state->dev_offset = static_cast<uint64_t>(0);
}
```
cuda rand， split时，需要不同的rank的local tensor的随机数生成时，按rank id做下偏移，以符合local拼接出来等于global的设定。猜测和这里的处理有关。

strint commented on oneflow/user/kernels/distributions/uniform_int_distribution.cu:
```diff
@@ -32,10 +33,13 @@ __device__ int64_t GenUniformInt(curandState* state, const int64_t low, const in
   curandState localState = state[id];
+  CUDA_1D_KERNEL_LOOP(i, elem_cnt * rank_id) {
+    static_cast<T>(GenUniformInt(&localState, low, high));
+  }
```
这里也增加下注释

grybd replied:
ok

hjchen2 commented on oneflow/user/kernels/distributions/uniform_int_distribution.cpp:
```diff
@@ -39,6 +40,8 @@ void UniformIntDistribution<DeviceType::kCPU, T>::operator()(
   // std::uniform_int_distribution generates [low, high], but we want [low, high) here
   CPUUniformIntDistributionImpl<T> impl(low_, high_ - 1);
+  int64_t rank_id = GlobalProcessCtx::Rank();
+  for (int64_t i = 0; i < elem_cnt * rank_id; ++i) { impl(gen->engine()); }
```
这里只能对Split和Partial做跳过处理，如果是Broadcast是不能跳过的

grybd replied:
> 这里只能对Split和Partial做跳过处理，如果是Broadcast是不能跳过的

确实是这样，这里传入sbp信息做判断如果是spb=broadcast 不做偏移是否可以呢?

grybd replied:
@strint

grybd replied:
> 这里需要增加些注释

好的

hjchen2 commented on oneflow/user/kernels/distributions/uniform_int_distribution.cpp:
```diff
@@ -33,18 +34,26 @@ class CPUUniformIntDistributionImpl {
+    that the local spliced tensor is equal to the global tensor */
+    for (int64_t i = 0; i < elem_cnt * rank_id; ++i) { impl(gen->engine()); }
+    for (int64_t i = 0; i < elem_cnt; ++i) { dptr[i] = impl(gen->engine()); }
+  }else{
```
把else去掉，同时把`for (int64_t i = 0; i < elem_cnt; ++i) { dptr[i] = impl(gen->engine()); }`提到外面去

grybd replied:
ok

hjchen2 commented on oneflow/user/kernels/distributions/uniform_int_distribution.cu:
```diff
@@ -32,34 +33,46 @@ __device__ int64_t GenUniformInt(curandState* state, const int64_t low, const in
+  /*when sbp=s, offset according to rank id to conform to the setting\
+   that the local spliced tensor is equal to the global tensor */
+  CUDA_1D_KERNEL_LOOP(i, elem_cnt*rank_id) {
+    static_cast<T>(GenUniformInt(&localState, low, high));
```
去掉static_cast<T>

grybd replied:
ok

grybd replied:
ok

liujuncheng commented on oneflow/user/kernels/distributions/uniform_int_distribution.cu:
```diff
@@ -32,34 +33,46 @@ __device__ int64_t GenUniformInt(curandState* state, const int64_t low, const in
+  if(elem_cnt != cnt){
+  /*when sbp=s, offset according to rank id to conform to the setting\
+   that the local spliced tensor is equal to the global tensor */
+  CUDA_1D_KERNEL_LOOP(i, elem_cnt*rank_id) {
+    static_cast<T>(GenUniformInt(&localState, low, high));
+  }
```
1. 这个处理在rank数量很多的时候会很慢，这里其实应该可以使用 skipahead 达到同样的目的，开销是 O(1) 的
2. 如果这个 curandState 还要继续使用，那么在生成　elem_cnt 个元素之后，应该还有一个 skipahead
3. 这个写法只对S(0)有效，对S(1)是不是还是不满足一致性的要求

liujuncheng replied:
> 1. 这个处理在rank数量很多的时候会很慢，这里其实应该可以使用 skipahead 达到同样的目的，开销是 O(1) 的
> 2. 如果这个 curandState 还要继续使用，那么在生成　elem_cnt 个元素之后，应该还有一个 skipahead
> 3. 这个写法只对S(0)有效，对S(1)是不是还是不满足一致性的要求

使用 Philox4_32_10 生成器的 skipahead 看起来开销比较低，默认的生成器开销更高

grybd replied:
好的，我调研一下 skipahead 的实现，这样处理好像是会使得各rank经过S(0)拼接出来的tensor和各个rank经过S(1)后拼接出来的global tensor是不一致的，我再确认一下 @liujuncheng

liujuncheng replied:
> 



> 好的，我调研一下 skipahead 的实现，这样处理好像是会使得各rank经过S(0)拼接出来的tensor和各个rank经过S(1)后拼接出来的global tensor是不一致的，我再确认一下 @liujuncheng

skipahead 没有对应的文档，这个函数在 cuda 的 include 文件 curand_kernel.h 里面，我说的其他内容也是包含在这个文件里。

grybd replied:
好的，知道了

hjchen2 commented on oneflow/user/kernels/distributions/uniform_int_distribution.cpp:
```diff
@@ -33,18 +34,26 @@ class CPUUniformIntDistributionImpl {
+    ep::Stream* stream, const int64_t elem_cnt,const int64_t cnt, T* dptr,
     const std::shared_ptr<one::Generator>& generator) const {
   CHECK_GE(elem_cnt, 0);
   auto gen = CHECK_JUST(generator->Get<one::CPUGeneratorImpl>());
```
可以考虑给generator统一加一个discard接口，执行z次但不生成任何随机数。类似这里提到的https://www.cplusplus.com/reference/random/mersenne_twister_engine/discard/

grybd replied:
好的，我了解一下~

grybd commented:
randint op 在graph和eager模式下 S0、S1拼接的global tensor不一致
```python

def _test_consistent_rand_graph(low, high, shape,generator, placement):
    class ConsistentRandGraph(flow.nn.Graph):
        def __init__(self,):
            super().__init__()

        def build(self):
            #x = flow.randint(low, high, shape, placement=placement,generator=generator, sbp=flow.sbp.broadcast)
            y = flow.randint(low,high,shape,placement=placement,generator=generator,sbp=flow.sbp.split(0))
            
            #x = flow.randperm(10, dtype=flow.int64, placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.broadcast)
            #y = flow.randperm(10, dtype=flow.int64, placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.split(0))
            return y

    c_r_g = ConsistentRandGraph()
    res = c_r_g()
    print("-----------------------------------------------------------")
    print("y: ",res)
    print("y.to_local(): ",res.to_local())

if __name__ == "__main__":
    generator = flow.Generator() 
    generator.manual_seed(0)
    _test_consistent_rand_graph(1,10 ,(2,2), generator, placement=flow.env.all_device_placement("cuda"))
    print("============================================================")
    print("eager mode: ",flow.randint(1,10,(2,2),generator=generator, placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.split(0)))
    print("============================================================")
 ```
T0 输出:
```shell 
-----------------------------------------------------------
y:  tensor([[5, 2],
        [5, 2]], placement=oneflow.placement(device_type="cuda", machine_device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
y.to_local():  tensor([[5, 2]], device='cuda:0', dtype=oneflow.int64)
============================================================
eager mode:  tensor([[5, 2],
        [5, 2]], placement=oneflow.placement(device_type="cuda", machine_device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
============================================================
```

当把S0改为S1
```python
def _test_consistent_rand_graph(low, high, shape,generator, placement):
    class ConsistentRandGraph(flow.nn.Graph):
        def __init__(self,):
            super().__init__()

        def build(self):
            #x = flow.randint(low, high, shape, placement=placement,generator=generator, sbp=flow.sbp.broadcast)
            y = flow.randint(low,high,shape,placement=placement,generator=generator,sbp=flow.sbp.split(1))
            
            #x = flow.randperm(10, dtype=flow.int64, placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.broadcast)
            #y = flow.randperm(10, dtype=flow.int64, placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.split(0))
            return y

    c_r_g = ConsistentRandGraph()
    res = c_r_g()
    print("-----------------------------------------------------------")
    print("y: ",res)
    print("y.to_local(): ",res.to_local())

if __name__ == "__main__":
    generator = flow.Generator() 
    generator.manual_seed(0)
    _test_consistent_rand_graph(1,10 ,(2,2), generator, placement=flow.env.all_device_placement("cuda"))
    print("============================================================")
    print("eager mode: ",flow.randint(1,10,(2,2),generator=generator, 
    placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.split(1)))
    print("============================================================")

```
T0输出：
```shell
-----------------------------------------------------------
y:  tensor([[5, 5],
        [2, 2]], placement=oneflow.placement(device_type="cuda", machine_device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=1),), dtype=oneflow.int64)
y.to_local():  tensor([[5],
        [2]], device='cuda:0', dtype=oneflow.int64)
============================================================
eager mode:  tensor([[5, 5],
        [2, 2]], placement=oneflow.placement(device_type="cuda", machine_device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=1),), dtype=oneflow.int64)
============================================================
```

可以看到graph的输出结果 S0：y:  tensor([[5, 2], 
        [5, 2]]    和 S1：y:  tensor([[5, 5],
        [2, 2]]    不一致 ，eager模式下也会出现同样情况

grybd commented:
@strint 有空帮忙review 一下哈， 提交的代码为什么在eager模式会出现 Segmentation fault (core dumped) 问题呢
示例：
```python
import numpy as np
import oneflow as flow
import torch
"""
class TestGraph(flow.nn.Graph):
    def __init__(self):
        super().__init__()

    def build(self):
        output = flow.arange(start=1,end=10,step=1,placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.split(0))
        return output

test = TestGraph()
out = test()
print("xxx: ",out)
print("out_1: ",out.to_local())
print("Not in graph: ",flow.arange(start=1,end=10,step=1,placement=flow.env.all_device_placement("cuda"),sbp=flow.sbp.split(0)))
placement=flow.env.all_device_placement("cuda")
sbp=flow.sbp.split(0)
"""

def _test_consistent_rand_graph(low, high, shape,generator, placement):
    class ConsistentRandGraph(flow.nn.Graph):
        def __init__(self,):
            super().__init__()

        def build(self):
            x = flow.randint(low, high, shape, placement=placement,generator=generator, sbp=flow.sbp.split(0))
            y = flow.randint(low,high,shape,placement=placement,generator=generator,sbp=flow.sbp.split(1))
            return [x,y]

    c_r_g = ConsistentRandGraph()
    res = c_r_g()
    print("-----------------------------------------------------------")
    print("after S0 global tensor: \n",res[0])
    print("after S0 local tensor: \n ",res[0].to_local())
    print("-----------------------------------------------------------")
    print("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
    print("after S1 global tensor: \n",res[1])
    print("after S1 local tensor: \n ",res[1].to_local())
    print("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")
  

if __name__ == "__main__":
    generator = flow.Generator() 
    generator.manual_seed(0)
    _test_consistent_rand_graph(1,10 ,(4,4), generator, placement=flow.env.all_device_placement("cpu"))
    print("============================================================")
    print("eager mode after S0: \n",flow.randint(1,10,(4,4),generator=generator, placement=flow.env.all_device_placement("cpu"),sbp=flow.sbp.split(0)))
    generator = flow.Generator() 
    generator.manual_seed(0)
    print("eager mode after S1: \n",flow.randint(1,10,(4,4),generator=generator, placement=flow.env.all_device_placement("cpu"),sbp=flow.sbp.split(1)))
    print("============================================================")
    ```
在T0上输出的信息：
```shell
-----------------------------------------------------------
after S0 global tensor: 
 tensor([[5, 6, 7, 8],
        [6, 8, 5, 8],
        [4, 6, 6, 4],
        [4, 3, 9, 1]],
       placement=oneflow.placement(device_type="cpu", machine_device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=0),), dtype=oneflow.int64)
after S0 local tensor: 
  tensor([[5, 6, 7, 8],
        [6, 8, 5, 8]], dtype=oneflow.int64)
-----------------------------------------------------------
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
after S1 global tensor: 
 tensor([[5, 6, 4, 6],
        [7, 8, 6, 4],
        [6, 8, 4, 3],
        [5, 8, 9, 1]],
       placement=oneflow.placement(device_type="cpu", machine_device_ids={0 : [0, 1]}, hierarchy=(2,)),
       sbp=(oneflow.sbp.split(axis=1),), dtype=oneflow.int64)
after S1 local tensor: 
  tensor([[5, 6],
        [7, 8],
        [6, 8],
        [5, 8]], dtype=oneflow.int64)
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
============================================================
eager mode after S0: 
Segmentation fault (core dumped) ```
